### PR TITLE
Analytic UI updates

### PR DESCRIPTION
**Pull request by n288TJYRX**

### CODE REVIEW

File: analytics/analytics-ui/src/app/parameters/parameters.module.ts [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+import { FlexLayoutModule } from '@angular/flex-layout';
+import { MAT_DATE_LOCALE } from '@angular/material/core';
+
+@NgModule({
+  declarations: [ParameterInputComponent, ParameterFormComponent, OperationComponent],
+  imports: [
+    CommonModule, MaterialModule, FormsModule, ReactiveFormsModule, FlexLayoutModule
+  ],
+  providers: [
+    {provide: MAT_DATE_LOCALE, useValue: 'en-CA'},
```
Comment (d47853) [inline analytics/analytics-ui/src/app/parameters/parameters.module.ts]: Shouldn't this just use the local users locale?

File: analytics/analytics-ui/src/app/results/html/html.component.ts [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+})
+
+export class HtmlComponent implements AfterViewInit {
+
+  constructor(
+    private results: ResultsService,
+  ) { }
+
+  ngAfterViewInit() {
+    const html = this.results.get();
```
Comment (d47853) [inline analytics/analytics-ui/src/app/results/html/html.component.ts]: What happens when there has been no results yet?

File: analytics/analytics-ui/src/app/results/results.component.html [hunk 3]
```diff
@@ -0,0 +1,17 @@
+<div style="margin-left:110px">
```
Comment (d47853) [inline analytics/analytics-ui/src/app/results/results.component.html]: Move styling into css file

File: analytics/analytics-ui/src/app/results/results.component.html [hunk 4]
```diff
@@ -0,0 +1,17 @@
...
+
+  <!--Title-->
+  <h1>Results</h1>
+
+  <!--Back button-->
+  <button mat-raised-button color="secondary" id="execute-analytic" (click)="goback()">BACK
```
Comment (d47853) [inline analytics/analytics-ui/src/app/results/results.component.html]: This should be a link, not a button. I don't think you really need a newline in the button

File: analytics/analytics-ui/src/app/results/results.component.ts [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+  outputType;
+  selected;
+
+  constructor(
+    private analyticsService: AnalyticsService,
+    private location: Location
+  ) { }
+
+  goback() {
+    this.location.back();
```
Comment (d47853) [inline analytics/analytics-ui/src/app/results/results.component.ts]: see link comment in html

File: analytics/analytics-ui/src/app/results/table/table.component.html [hunk 6]
```diff
@@ -0,0 +1,35 @@
+<!--Column Remove Selector and Button-->
+<div>
+  <mat-form-field style="margin-left:110px">
```
Comment (d47853) [inline analytics/analytics-ui/src/app/results/table/table.component.html]: Add this styling to the table css file

File: analytics/analytics-ui/src/app/results/table/table.component.html [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+    <mat-label>Columns</mat-label>
+    <mat-select [(value)]="selected">
+      <mat-option *ngFor="let column of columnsToDisplay" value={{column}}>
+        {{ column | uppercase }}
```
Comment (d47853) [inline analytics/analytics-ui/src/app/results/table/table.component.html]: Is there a need to uppercase the values? Generally people find it harder to scan uppercase text.

File: analytics/analytics-ui/src/app/results/table/table.component.html [hunk 8]
```diff
@@ -0,0 +1,35 @@
...
+      </mat-option>
+    </mat-select>
+  </mat-form-field>
+  <button style="margin-left:30px" mat-raised-button (click)="removeColumn()"> Remove
```
Comment (d47853) [inline analytics/analytics-ui/src/app/results/table/table.component.html]: Add to css file

File: analytics/analytics-ui/src/app/results/table/table.component.html [hunk 9]
```diff
@@ -0,0 +1,35 @@
...
+    column </button>
+</div>
+
+<!--Data Table-->
+<mat-card>
+  <div class="example-container">
+    <mat-table [dataSource]="data.results" matSort>
+      <ng-container [matColumnDef]="column" *ngFor="let column of columnsToDisplay">
+        <mat-header-cell *matHeaderCellDef mat-sort-header>
+          <h3>{{ column | uppercase }}</h3>
```
Comment (d47853) [inline analytics/analytics-ui/src/app/results/table/table.component.html]: See above comment on uppercasing

File: analytics/analytics-ui/src/app/results/table/table.component.css [hunk 10]
```diff
@@ -0,0 +1 @@
+
```
Comment (d47853) [inline analytics/analytics-ui/src/app/results/table/table.component.css]: Remove any empty css files

File: analytics/analytics-ui/src/app/results/table/table.component.spec.ts [hunk 11]
```diff
@@ -0,0 +1,112 @@
+import { async, ComponentFixture, TestBed } from '@angular/core/testing';
```
Comment (d47853) [inline analytics/analytics-ui/src/app/results/table/table.component.spec.ts]: More testing needed here

File: analytics/analytics-ui/src/app/services/results.service.ts [hunk 12]
```diff
@@ -31,7 +31,7 @@ export class ResultsService {
 
   /** Update the results */
   update = function(newResults) {
-    if (newResults != null && newResults !== '') {
+    if (newResults !== undefined && newResults !== null && newResults !== '') {
```
Comment (d47853) [inline analytics/analytics-ui/src/app/services/results.service.ts]:
```suggestion
    if (newResults != null && newResults !== '') {
```

File: analytics/analytics-ui/src/app/services/type.service.ts [hunk 13]
```diff
@@ -0,0 +1,193 @@
+/*
```
Comment (d47853) [inline analytics/analytics-ui/src/app/services/type.service.ts]: Remove this file as it isn't used anywhere in the app

File: analytics/analytics-ui/src/styles.css [hunk 14]
```diff
@@ -39,26 +36,6 @@ div {
...
-  justify-content: flex-start;
-  flex-wrap: wrap;
-}
-
-mat-card-content {
-  flex: 1;
-  padding: 0.25em 0.3em 0;
-}
-
 mat-card {
```
Comment (d47853) [inline analytics/analytics-ui/src/styles.css]: This should probably be on a per component basis

File: analytics/analytics-ui/src/styles.css [hunk 15]
```diff
@@ -72,497 +49,20 @@ mat-card {
...
-  z-index: 2;
-}
-
-app-analytic {
-  width: 100%;
-  height: 100%;
-}
-
-app-analytic > mat-card {
+app-analytic>mat-card {
```
Comment (d47853) [inline analytics/analytics-ui/src/styles.css]: Move into the analytic component css

File: analytics/analytics-ui/src/styles.css [hunk 16]
```diff
@@ -72,497 +49,20 @@ mat-card {
...
-
-.page {
-  text-align: center;
-}
-
-.routeActive {
-  font-weight: bold;
-}
-
 .example-container {
```
Comment (d47853) [inline analytics/analytics-ui/src/styles.css]: This should be moved to where ever it's used

File: analytics/analytics-ui/src/styles.css [hunk 17]
```diff
@@ -72,497 +49,20 @@ mat-card {
...
-  background: white;
-  min-width: 2000px;
-}
-
-.mat-header-row {
-  position: sticky;
-  top: 0;
-}
-
 .iconContainer {
```
Comment (d47853) [inline analytics/analytics-ui/src/styles.css]: Move to where ever it's used. It shouldn't be global

File: pom.xml [hunk 18]
```diff
@@ -21,7 +21,7 @@
     <parent>
         <groupId>uk.gov.gchq.gaffer</groupId>
         <artifactId>gaffer2</artifactId>
-        <version>1.10.0-RC2</version>
+        <version>1.10.1-SNAPSHOT</version>
```
Comment (d47853) [inline pom.xml]: Undo this change and merge develop into your branch to avoid conflicts.

File: analytics/analytics-ui/src/app/results/html/html.component.ts (hunk 2 above)
Comment (n288TJYRX) [reply analytics/analytics-ui/src/app/results/html/html.component.ts]: If you have a look at analytics.service.ts:142 we don't load the results page until the results have loaded. Until then you'll see a buffering spinner symbol once you click execute on the parameters page.

File: analytics/analytics-ui/src/app/results/html/html.component.spec.ts [hunk 19]
```diff
@@ -0,0 +1,70 @@
...
+    fixture = TestBed.createComponent(HtmlComponent);
+    component = fixture.componentInstance;
+    fixture.detectChanges();
+  });
+
+  it('should create', () => {
+    expect(component).toBeTruthy();
+  });
+
+  it('should equal', () => {
```
Comment (d47853) [inline analytics/analytics-ui/src/app/results/html/html.component.spec.ts]: Can this test name be improved?

File: analytics/analytics-ui/src/app/results/results.component.html [hunk 20]
```diff
@@ -0,0 +1,18 @@
+<div style="margin-left:110px">
+
+  <!--Title-->
+  <h1>Results</h1>
+
+  <!--Back button-->
+  <button mat-raised-button id="back-params" [routerLink]="['..']" style="cursor:pointer"
```
Comment (d47853) [inline analytics/analytics-ui/src/app/results/results.component.html]: You should use the <a> tag for links. https://material.angular.io/components/button/examples

File: analytics/analytics-ui/src/app/results/results.component.spec.ts [hunk 21]
```diff
@@ -0,0 +1,519 @@
...
+  columns: ['SOURCE', 'result type', 'value', 'GROUP', 'DESTINATION', 'DIRECTED', 'prop2', 'count', 'prop1'],
+  allColumns: ['SOURCE', 'result type', 'value', 'GROUP', 'DESTINATION', 'DIRECTED', 'prop2', 'count', 'prop1'],
+  groups: ['BasicEdge1', 'BasicEdge2', 'BasicEntity1', 'BasicEntity2', ''],
+  allGroups: ['BasicEdge1', 'BasicEdge2', 'BasicEntity1', 'BasicEntity2', ''],
+  types: ['Edge', 'Entity', 'String', 'Integer', 'EntitySeed'],
+  allTypes: ['Edge', 'Entity', 'String', 'Integer', 'EntitySeed'],
+  tooltips: {}
+};
+
+const types = {
```
Comment (d47853) [inline analytics/analytics-ui/src/app/results/results.component.spec.ts]: These are not required.

File: analytics/analytics-ui/src/app/results/table/table.component.html [hunk 22]
```diff
@@ -0,0 +1,35 @@
...
+<div>
+  <mat-form-field>
+    <mat-label>Columns</mat-label>
+    <mat-select [(value)]="selected">
+      <mat-option *ngFor="let column of columnsToDisplay" value={{column}}>
+        {{ column }}
+      </mat-option>
+    </mat-select>
+  </mat-form-field>
+  <button mat-raised-button style="margin-left: 30px;" (click)="removeColumn()" aria-label="Remove selected column">
```
Comment (d47853) [inline analytics/analytics-ui/src/app/results/table/table.component.html]: Please can you move styling into css file

File: analytics/analytics-ui/src/app/services/analytics.service.ts [hunk 23]
```diff
@@ -48,16 +50,19 @@ export class AnalyticsService {
...
-    // Convert to an integer
-    newValue = parseInt(newValue, 10);
+
     // Look for the parameter in the list of parameters and set the new current value
     for (const parameterPair of this.arrayAnalytic.uiMapping) {
-        if (parameterPair[0] === parameterName) {
-          parameterPair[1].currentValue = newValue;
-          return;
-        }
+      if (parameterPair[0] === parameterName) {
```
Comment (d47853) [inline analytics/analytics-ui/src/app/services/analytics.service.ts]: This feels quite messy. Is there a way we could tidy this up?

File: analytics/analytics-ui/src/app/services/filter.pipe.ts [hunk 24]
```diff
@@ -0,0 +1,14 @@
+import { Pipe, PipeTransform } from '@angular/core';
+@Pipe({
+    name: 'filter'
+})
+export class FilterPipe implements PipeTransform {
+    transform(items: any[], searchText: string): any[] {
```
Comment (d47853) [inline analytics/analytics-ui/src/app/services/filter.pipe.ts]: It seems dangerous to use any[] here. You should make use of typescript and create an Analytic class. Also, you should consider renaming this to 'analytic-filter' as it's specific to analytics.

File: analytics/analytics-ui/src/app/results/results.component.spec.ts [hunk 25]
```diff
@@ -0,0 +1,251 @@
...
+import { ErrorService } from '../services/error.service';
+import { TableComponent } from './table/table.component';
+import { HtmlComponent } from './html/html.component';
+
+const routes: Routes = [
+  { path: '**', redirectTo: 'analytics' }
+];
+
+const fullResultsData = [
+  {
```
Comment (d47853) [inline analytics/analytics-ui/src/app/results/results.component.spec.ts]: Are either of these values actually used in this test?

File: analytics/analytics-ui/src/app/services/analytic-filter.pipe.spec.ts [hunk 26]
```diff
@@ -0,0 +1,58 @@
+import { AnalyticFilterPipe } from './analytic-filter.pipe';
```
Comment (d47853) [inline analytics/analytics-ui/src/app/services/analytic-filter.pipe.spec.ts]: Can you add a copyright header here

File: analytics/analytics-ui/src/app/services/analytic-filter.pipe.ts [hunk 27]
```diff
@@ -0,0 +1,18 @@
+import { Pipe, PipeTransform } from '@angular/core';
```
Comment (d47853) [inline analytics/analytics-ui/src/app/services/analytic-filter.pipe.ts]: Copyright header is needed for this method. Perhaps you could add something which automatically checks this.

File: analytics/analytics-ui/src/app/results/table/table.component.ts [hunk 28]
```diff
@@ -0,0 +1,75 @@
+import { Component, OnInit, Input, ViewChild, AfterViewInit } from '@angular/core';
```
Comment (d47853) [inline analytics/analytics-ui/src/app/results/table/table.component.ts]: Can you add a copyright header here.

File: analytics/analytics-ui/src/app/services/analytic-filter.pipe.spec.ts (hunk 26 above)
Comment (d47853) [inline analytics/analytics-ui/src/app/services/analytic-filter.pipe.spec.ts]: Can you add a copyright header here (and every .ts file where there isn't one)

File: analytics/analytics-ui/src/app/services/analytic-filter.pipe.ts [hunk 29]
```diff
@@ -0,0 +1,18 @@
...
+class Analytic {
```
Comment (d47853) [inline analytics/analytics-ui/src/app/services/analytic-filter.pipe.ts]: If this class was more representative of the AnalyticDetail class, it could be more widely used throughout the app (such as in the Analytics Service). This gives the benefit of type saftely and makes the code easier to maintain. Furthermore, it should be moved into it's own file.

File: analytics/analytics-ui/src/app/services/analytic-filter.pipe.ts (hunk 27 above)
Comment (d47853) [inline analytics/analytics-ui/src/app/services/analytic-filter.pipe.ts]: This filter should not be located in the services directory as it is only used in the analytics component and therefore shouldn't be imported into every component.

File: analytics/analytics-ui/src/app/results/table/table.component.html [hunk 30]
```diff
@@ -0,0 +1,35 @@
+<!--Column Remove Selector and Button-->
+<div>
+  <mat-form-field>
+    <mat-label>Columns</mat-label>
+    <mat-select [(value)]="selected">
+      <mat-option *ngFor="let column of columnsToDisplay" value={{column}}>
```
Comment (d47853) [inline analytics/analytics-ui/src/app/results/table/table.component.html]: It seems like there are a lot of button clicks needed to remove multiple columns. Perhaps providing a multi-select dropdown would make for a better user experience - similar in design to the Gaffer UI. It looks like the removeColumn() may need a rework too but I'll put more detail on that in the .ts file.

File: analytics/analytics-ui/src/app/results/table/table.component.ts [hunk 31]
```diff
@@ -0,0 +1,75 @@
...
+    this.data.results.sort = this.sort;
+  }
+
+  ngAfterViewInit() {
+    this.data.results.paginator = this.paginator;
+    this.data.results.sort = this.sort;
+  }
+
+  removeColumn() {
+    Object.keys(this.columnsToDisplay).forEach(key => {
```
Comment (d47853) [inline analytics/analytics-ui/src/app/results/table/table.component.ts]: This seems really convoluted and it looks like you're changing the array as you iterate over it which can lead to unexpected behaviour or errors. What might be better is if you use the multiple select, you can have two models, one for the available columns (defined by the results returned) and one for the columns the user wants to see.

File: analytics/analytics-ui/src/app/services/analytics.service.ts [hunk 32]
```diff
@@ -0,0 +1,152 @@
...
+        analytic.uiMapping.get(parameterKey).currentValue = false;
+      } else {
+        analytic.uiMapping.get(parameterKey).currentValue = null;
+      }
+    }
+    this.analytic = analytic;
+  };
+
+  /** Execute the analytic operation */
+  executeAnalytic = function() {
```
Comment (d47853) [inline analytics/analytics-ui/src/app/services/analytics.service.ts]: This is quite a long method. Would you consider shortening it / breaking it up?

File: analytics/analytics-ui/src/app/services/analytics.service.ts [hunk 33]
```diff
@@ -0,0 +1,152 @@
...
+
+import { startsWith } from 'lodash';
+
+const OPERATION_CHAIN_CLASS = 'uk.gov.gchq.gaffer.operation.OperationChain';
+const MAP_OPERATION_CLASS = 'uk.gov.gchq.gaffer.operation.impl.Map';
+
+// Used to store and get the selected analytic
+@Injectable()
+export class AnalyticsService {
+  analytic: Analytic; // The selected analytic
```
Comment (d47853) [inline analytics/analytics-ui/src/app/services/analytics.service.ts]: How much effort would there be to make this service stateless?

File: analytics/analytics-functions/src/main/java/uk/gov/gchq/gaffer/analytic/function/ToMap.java [hunk 34]
```diff
@@ -1,5 +1,5 @@
 /*
- * Copyright 2019 Crown Copyright
+ * Copyright 2019-2020-2020 Crown Copyright
```
Comment (d47853) [inline analytics/analytics-functions/src/main/java/uk/gov/gchq/gaffer/analytic/function/ToMap.java]: This should just be 2019-2020 if it's an old file and 2020 if it's a new file
